File: nim/time_stamp.nim
import time_nanos, strutils, parseutils

type
  Stamp* = distinct int64

# Bunch of methods we borrow to allow mixing stamps with seconds.
proc `+`*(x: Stamp, y: Nano): Stamp {.borrow.}
proc `+`*(x: Nano, y: Stamp): Stamp {.borrow.}
proc `-`*(x: Stamp, y: Nano): Stamp {.borrow.}
proc `-`*(x: Nano, y: Stamp): Stamp {.borrow.}

const
  year_start = "".len
  month_start = "YYYY-".len
  days_start = "YYYY-MM-".len
  hours_start = "YYYY-MM-DDT".len
  minutes_start = "YYYY-MM-DDThh:".len
  seconds_start = "YYYY-MM-DDThh:mm:".len
  nanos_start = "YYYY-MM-DDThh:mm:ss:".len
  max_stamp_len = "YYYY-MM-DDThh:mm:ss:012345678".len
  epoch_offset = 1970
  days_in_a_month = 30 # Because I'm worth it…


proc year*(x: Stamp): int {.procvar.} = Nano(x).year + epoch_offset
proc week*(x: Stamp): int {.procvar.} = Nano(x).week
proc month*(x: Stamp): int {.procvar.} = Nano(x).month
proc day*(x: Stamp): int {.procvar.} = Nano(x).day
proc hour*(x: Stamp): int {.procvar.} = Nano(x).hour
proc minute*(x: Stamp): int {.procvar.} = Nano(x).minute
proc second*(x: Stamp): int {.procvar.} = Nano(x).second
proc microsecond*(x: Stamp): int {.procvar.} = Nano(x).microsecond
proc millisecond*(x: Stamp): int {.procvar.} = Nano(x).millisecond
proc nanosecond*(x: Stamp): int {.procvar.} = Nano(x).nanosecond


proc date*(x: string): Stamp =
  ## Converts a random string into a timestamp value.
  ##
  ## For simplicity this is just a naive implementation that will easily break
  ## under the weight of any real user input, there is no real user friendly
  ## validation here. We expect strings to be in the format YYYY-MM-DD and
  ## that's it. Ok, maybe we allow times later too…
  assert x.len >= 10 and x.len <= max_stamp_len
  var token: string
  do_assert 4 == x.parse_while(token, Digits, year_start)
  var
    yyyy: int
    mm: int
    dd: int
  do_assert 4 == token.parse_int(yyyy)
  do_assert yyyy >= epoch_offset

  do_assert 2 == x.parse_while(token, Digits, month_start)
  do_assert 2 == token.parse_int(mm)
  do_assert mm > 0 and mm < 13

  do_assert 2 == x.parse_while(token, Digits, days_start)
  do_assert 2 == token.parse_int(dd)
  do_assert dd > 0 and dd < 32

  # Finally, convert the individual values to a (fake) calendar.
  result = Stamp((yyyy - epoch_offset) * u_year +
    (mm - 1) * days_in_a_month * u_day + (dd - 1) * u_day)

  if x.len < minutes_start - 1:
    return

  # Ugh, we got some times… ok, try to parse them.
  var hh: int
  do_assert 2 == x.parse_while(token, Digits, hours_start)
  do_assert 2 == token.parse_int(hh)
  do_assert hh >= 0 and hh < 24
  result = result + hh.h

  if x.len < seconds_start - 1:
    return

  var minutes: int
  do_assert 2 == x.parse_while(token, Digits, minutes_start)
  do_assert 2 == token.parse_int(minutes)
  do_assert minutes >= 0 and minutes < 60
  result = result + minutes.i

  if x.len < nanos_start - 1:
    return

  var seconds: int
  do_assert 2 == x.parse_while(token, Digits, seconds_start)
  do_assert 2 == token.parse_int(seconds)
  do_assert seconds >= 0 and seconds < 60
  result = result + seconds.s

  if x.len > nanos_start:
    var nanos: int
    var num_nanos = x.parse_while(token, Digits, nanos_start)
    do_assert num_nanos > 0 and num_nanos < 10
    do_assert num_nanos == token.parse_int(nanos)
    # Compensate lack of decimals during parsing with pow-like loop.
    while num_nanos < 9:
      nanos = nanos * 10
      num_nanos.inc
    result = result + Nano(nanos)


proc d*(x: string): Stamp {.inline.} = x.date
proc zeroalign(x: string, count = 2): string {.inline.} = align(x, count, '0')

proc `$`*(x: Stamp): string =
  ## Represents the time as a computer readable date.
  ##
  ## Because we really are not humans, are we? If we were I would have to deal
  ## with regions and locale and all that bulsh…KILL ALL HUMANS.
  var total = Nano(x)
  let seconds = total mod u_minute
  total = total - seconds

  let minutes = total mod u_hour
  total = total - minutes

  let hours = total mod u_day
  total = total - hours

  let
    days = total mod u_year
    years = (total - days) div u_year

  result = $(epoch_offset + years) & "."

  # Convert days to numbers for final in year calculations.
  var numeric_days = days div u_day
  let numeric_months = numeric_days div days_in_a_month
  numeric_days = numeric_days mod days_in_a_month

  result &= zeroalign($(1 + numeric_months)) & "." &
    zeroalign($(1 + numeric_days))

  let
    numeric_seconds = seconds div u_second
    numeric_minutes = minutes div u_minute
    numeric_hours = hours div u_hour
    numeric_nanos = int(seconds mod u_second)

  # Return already if the time ends at midnight.
  if numeric_seconds < 1 and numeric_minutes < 1 and
      numeric_hours < 1 and numeric_nanos < 1:
    return

  # Aww… format an hour then.
  result &= "T" & zeroalign($numeric_hours) &
    ":" & zeroalign($numeric_minutes) &
    ":" & zeroalign($numeric_seconds)

  if numeric_nanos > 0:
    result &= "." & zeroalign($numeric_nanos, 9)


# Helper procs to map lists.
proc `+`*(x: Stamp, y: seq[Nano]): seq[Stamp] =
  result.new_seq(y.len)
  var pos = 0
  while pos < y.len:
    result[pos] = x + y[pos]
    pos.inc


# Helper to apply functions to sequences using subscript syntax.
proc `[]`*[S,T](x: seq[S], filter: proc (x: S): T): seq[T] =
  result.new_seq(x.len)
  var pos = 0
  while pos < x.len:
    result[pos] = filter(x[pos])
    pos.inc


proc test_stamps*() =
  echo "Testing stamps\n"
  var a = date"2012-01-01"
  echo "let's start at ", a
  echo "plus one day is ", a + 1.d
  echo "plus one month is ", a + 1.m
  echo "plus one month and a day is ", a + 1.m + 1.d
  echo "…plus 1h15i17s ", a + 1.m + 1.d + 1.h + 15.i + 17.s
  echo "…plus 23 hours ", a + 1.m + 2.d - 1.h
  echo "2001.01.01T01".date
  echo "2001.01.01T02:01".date
  echo "2001.01.01T03:02:01".date
  echo "2001.01.01T04:09:02.1".date
  echo date"2001.01.01T04:09:02.12"
  echo "2001.01.01T04:09:02.123".date
  echo "2001.01.01T05:04:03.0123".date
  echo "2001.01.01T06:05:04.012345678".date
  a = "2001.01.01T06:05:04.012345678".date
  echo "\tyear ", a.year
  echo "\tmonth ", a.month
  echo "\tday ", a.day
  echo "\thour ", a.hour
  echo "\tminute ", a.minute
  echo "\tsecond ", a.second
  echo "\tmicrosecond ", a.microsecond
  echo "\tmillisecond ", a.millisecond
  echo "\tnanosecond ", a.nanosecond

File: cpp/time_stamp.cpp
#include "time_stamp.h"

#include "time_nanos.h"


using namespace std;


const int YEAR_START = 0;
const int MONTH_START = YEAR_START + 5;
const int DAYS_START = MONTH_START + 3;
const int HOURS_START = DAYS_START + 3;
const int MINUTES_START = HOURS_START + 3;
const int SECONDS_START = MINUTES_START + 3;
const int NANOS_START = SECONDS_START + 3;
const int MAX_STAMP_LEN = 30;
const int DAYS_IN_A_MONTH = 30;
const int EPOCH_OFFSET = 1970;
const long long ONE_SECOND = 1000000000;

ostream& operator<<(std::ostream& o, const Stamp& x)
{
	long long total = x.val;
	const long long seconds = total % u_minute.val;
	total = total - seconds;

	const long long minutes = total % u_hour.val;
	total = total - minutes;

	const long long hours = total % u_day.val;
	total = total - hours;

	const long long days = total % u_year.val;
	const long long years = (total - days) / u_year.val;

	o << (EPOCH_OFFSET + years) << "." << setfill('0') << right;
	long long numeric_days = days / u_day.val;
	const long long numeric_months = 1 + numeric_days / DAYS_IN_A_MONTH;
	numeric_days = 1 + (numeric_days % DAYS_IN_A_MONTH);

	o << setw(2) << numeric_months << '.';
	o << setw(2) << numeric_days;

	const long long numeric_seconds = seconds / u_second.val;
	const long long numeric_minutes = minutes / u_minute.val;
	const long long numeric_hours = hours / u_hour.val;
	const long long numeric_nanos = seconds % u_second.val;

	if (numeric_seconds < 1 && numeric_minutes < 1
			&& numeric_hours < 1 && numeric_nanos < 1)
		return o;

	o << 'T';
	o << setw(2) << numeric_hours << ':';
	o << setw(2) << numeric_minutes << ':';
	o << setw(2) << numeric_seconds;

	if (numeric_nanos > 0) {
		o << '.' << setw(9) << numeric_nanos;
	}
	return o;
}

//const auto a = "2012-0101"_date;

void test_stamp()
{
	cout << "Testing stamp module" << endl << endl;

	auto a = "2012-01-01"_date;
	cout << "let's start at " << a << endl;
	cout << "plus one day is " << a + 1_d << endl;
	cout << "plus one month is " << a + 1_m << endl;
	cout << "plus one month and a day is " << a + 1_m + 1_d << endl;
	cout << "…plus 1h15i17s " << a + 1_m + 1_d + 1_h + 15_i + 17_s << endl;
	cout << "…plus 23 hours " << a + 1_m + 2_d - 1_h << endl;
	cout << "2001.01.01T01"_date << endl;
	cout << "2001.01.01T02:01"_date << endl;
	cout << "2001.01.01T03:02:01"_date << endl;
	cout << "2001.01.01T04:09:02.1"_date << endl;
	cout << "2001.01.01T04:09:02.12"_date << endl;
	cout << "2001.01.01T04:09:02.123"_date << endl;
	cout << "2001.01.01T05:04:03.0123"_date << endl;
	cout << "2001.01.01T06:05:04.012345678"_date << endl;
	a = "2001.01.01T06:05:04.012345678"_date;
	cout << "\tyear " << a.year() << endl;
	cout << "\tmonth " << a.month() << endl;
	cout << "\tday " << a.day() << endl;
	cout << "\thour " << a.hour() << endl;
	cout << "\tminute " << a.minute() << endl;
	cout << "\tsecond " << a.second() << endl;
	cout << "\tmicrosecond " << a.microsecond() << endl;
	cout << "\tmillisecond " << a.millisecond() << endl;
	cout << "\tnanosecond " << a.nanosecond() << endl;
}
